### PR TITLE
Feature/bucket creation

### PR DESCRIPTION
**Pull request by shawn-hurley**

### CODE REVIEW

**Comment by shawn-hurley** [comment]
@alaypatel07 @shubham-pampattiwar @dymurray Please give me some feedback on the API. But this is rebase for the API rework.

File: api/v1alpha1/bucket_types.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BucketSpec   `json:"spec,omitempty"`
+	Status BucketStatus `json:"status,omitempty"`
+}
+
+type BucketSpec struct {
+	Name               string                      `json:"name"`
+	CreationSecret     corev1api.SecretKeySelector `json:"bucketCreationSecret"`
+	EnableSharedConfig bool                        `json:"enableSharedConfig,omitempty"`
```
Comment (alaypatel07) [inline api/v1alpha1/bucket_types.go:22]: IIUC it is a conscious choice to make this default to `false` and we do not anticipate ever turning this around to defaulting it to true?

If Im not right, we might want to consider using `*bool`

File: controllers/bsl.go [hunk 2]
```diff
@@ -33,34 +34,46 @@ func (r *DPAReconciler) ValidateBackupStorageLocations(log logr.Logger) (bool, e
...
-			if err != nil {
-				return false, err
+			if bslSpec.Bucket.Credential.LocalObjectReference.Name == "" {
+				return false, fmt.Errorf("must provide a valid credential secret name")
 			}
-		// case when no supporting cloud provider is configured in bsl
-		default:
-			return false, fmt.Errorf("no valid provider configured for one of the backupstoragelocations")
+		}
+		if bslSpec.Bucket != nil && bslSpec.Velero != nil {
```
Comment (alaypatel07) [inline controllers/bsl.go:73]: is there ever a case where both spec.Bucket and spec.Velero are nil that we want to handle here?

File: api/v1alpha1/bucket_types.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package v1alpha1
+
+import (
+	corev1api "k8s.io/api/core/v1"
```
Comment (alaypatel07) [inline api/v1alpha1/bucket_types.go:4]: nit: in other places, we have used `corev1` might want to change this to make it consistent?

**Review by alaypatel07 — COMMENTED** [review]
Overall this looks good to me. Added a couple of questions!

File: api/v1alpha1/bucket_types.go (hunk 1 above)
Comment (shawn-hurley) [reply api/v1alpha1/bucket_types.go:22]: For future proof might as well do *bool +1

File: controllers/bsl.go (hunk 2 above)
Comment (shawn-hurley) [reply controllers/bsl.go:73]: I believe that this is covered in the ensureMapping function here: https://github.com/openshift/oadp-operator/pull/432/files#diff-92d442dfa6f9e391ba11832929f440817bcabe3f794700a7e9ecd9a10c309d24R293-R294

File: api/v1alpha1/bucket_types.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+type Bucket struct {
```
Comment (dymurray) [inline api/v1alpha1/bucket_types.go:11]: I don't like the name `DataProtectionBucket`, but wondering if this `Kind` gets confusing with other CRDs from ODF like 
```
bucketclasses.noobaa.io                                          
objectbucketclaims.objectbucket.io                          
objectbuckets.objectbucket.io                         
```

I think its fine... just wanted to bring it up.

File: api/v1alpha1/bucket_types.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BucketSpec   `json:"spec,omitempty"`
+	Status BucketStatus `json:"status,omitempty"`
+}
+
+type BucketSpec struct {
+	Name               string                   `json:"name"`
+	CreationSecret     corev1.SecretKeySelector `json:"bucketCreationSecret"`
```
Comment (dymurray) [inline api/v1alpha1/bucket_types.go:21]: NIT: `bucketCreationSecret` is overdescriptive on the JSON tag. Just `creationSecret`

File: api/v1alpha1/bucket_types.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	EnableSharedConfig *bool                    `json:"enableSharedConfig,omitempty"`
+	Tags               map[string]string        `json:"bucketTags,omitempty"`
```
Comment (dymurray) [inline api/v1alpha1/bucket_types.go:23]: NIT: same here, just `tags` for JSON

File: api/v1alpha1/bucket_types.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+type Bucket struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BucketSpec   `json:"spec,omitempty"`
+	Status BucketStatus `json:"status,omitempty"`
+}
+
+type BucketSpec struct {
```
Comment (dymurray) [inline api/v1alpha1/bucket_types.go:19]: NIT: Please add comments for each field if possible

File: api/v1alpha1/bucket_types.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	Name               string                   `json:"name"`
+	CreationSecret     corev1.SecretKeySelector `json:"bucketCreationSecret"`
+	EnableSharedConfig *bool                    `json:"enableSharedConfig,omitempty"`
+	Tags               map[string]string        `json:"bucketTags,omitempty"`
+	Region             string                   `json:"region"`
+}
+
+type BucketStatus struct {
+	Name              string       `json:"name"`
+	LastSyncTimestamp *metav1.Time `json:"lastSyncTimestamp,omitempty"`
```
Comment (dymurray) [inline api/v1alpha1/bucket_types.go:29]: NIT: `LastSynced`?

File: pkg/bucket/client.go [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+	s3Client, err := a.getS3Client()
+	if err != nil {
+		return false, err
+	}
+	createBucketInput := &s3.CreateBucketInput{
+		ACL:    aws.String(s3.BucketCannedACLPrivate),
+		Bucket: aws.String(a.bucket.Spec.Name),
+	}
+	if a.bucket.Spec.Region != "us-east-1" {
+		createBucketConfiguation := &s3.CreateBucketConfiguration{
```
Comment (dymurray) [inline pkg/bucket/client.go:91]: NIT: `createBucketConfiguration`

File: pkg/bucket/client.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+func (a awsBucketClient) Create() (bool, error) {
+	s3Client, err := a.getS3Client()
+	if err != nil {
+		return false, err
+	}
+	createBucketInput := &s3.CreateBucketInput{
+		ACL:    aws.String(s3.BucketCannedACLPrivate),
+		Bucket: aws.String(a.bucket.Spec.Name),
+	}
+	if a.bucket.Spec.Region != "us-east-1" {
```
Comment (dymurray) [inline pkg/bucket/client.go:90]: Assuming this is because this is the default region?

**Comment by shawn-hurley** [comment]
@dymurray because of the squash merge, I have had to do some nasty rebase stuff with commits sorry.

File: pkg/bucket/client.go (hunk 10 above)
Comment (shawn-hurley) [reply pkg/bucket/client.go:90]: Resolving this thread

File: api/v1alpha1/bucket_types.go (hunk 4 above)
Comment (shawn-hurley) [reply api/v1alpha1/bucket_types.go:11]: What about CloudBucket or something like  BackupStorageBucket?

File: api/v1alpha1/bucket_types.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BucketSpec   `json:"spec,omitempty"`
+	Status BucketStatus `json:"status,omitempty"`
+}
+
+type BucketSpec struct {
+	// Name is the name requested for the bucket
+	Name string `json:"name"`
+	// CreationSecret is the secret that is needed to be used while creating the bucket.
```
Comment (shubham-pampattiwar) [inline api/v1alpha1/bucket_types.go:22]: @shawn-hurley Would it make sense to have `provider` as a spec ? (eg: AWS/GCP/Noobaa) or add the `provider` spec as a part of `BucketBackupLocation` ?

File: api/v1alpha1/oadp_types.go [hunk 12]
```diff
@@ -108,10 +108,32 @@ type ApplicationConfig struct {
...
+	// +optional
+	// +nullable
+	BackupSyncPeriod *metav1.Duration `json:"backupSyncPeriod,omitempty"`
+}
+
 // BackupLocation defines the configuration for the DPA backup storage
 type BackupLocation struct {
 	// TODO: Add name/annotations/labels support
 	Velero *velero.BackupStorageLocationSpec `json:"velero"`
+	Bucket *BucketBackupLocation             `json:"bucket"`
```
Comment (shubham-pampattiwar) [inline api/v1alpha1/oadp_types.go:136]: Shouldn't we keep the `Bucket` as an optional spec ?

File: api/v1alpha1/bucket_types.go (hunk 11 above)
Comment (shawn-hurley) [reply api/v1alpha1/bucket_types.go:22]: I think it totally makes sense to have a provider as a spec value 🤦‍♂️

File: api/v1alpha1/oadp_types.go (hunk 12 above)
Comment (shawn-hurley) [reply api/v1alpha1/oadp_types.go:136]: Do you mean just add the omit empty or the comment??

Comment (shubham-pampattiwar) [reply api/v1alpha1/oadp_types.go:136]: yes, `omitempty` and `+optional` kubebuilder annotation.

File: api/v1alpha1/bucket_types.go (hunk 4 above)
Comment (dymurray) [reply api/v1alpha1/bucket_types.go:11]: Yeah both are good with me. I also don't hate Bucket :D

**Review by dymurray — APPROVED** [review]

**Comment by dymurray** [comment]
I have run e2e tests locally multiple times and all is working. I am merging to unblock ability to cut release. If this screws things up I take responsibility and will revert/fix :) :boom:

**Pull request merged**